gufe/base.py

```diff
@@ -0,0 +1,381 @@
+            and '__module__' in dct)
+
+
+def is_gufe_key_dict(dct: Any):
+    return isinstance(dct, dict) and ":gufe-key:" in dct
+
+
+# conveniences to get a class from module/class name
+def import_qualname(modname: str, qualname: str, remappings=REMAPPED_CLASSES):
+    if (modname, qualname) in remappings:
+        modname, qualname = remappings[(modname, qualname)]
+
+    result = importlib.import_module(modname)
+    for name in qualname.split('.'):
+        result = getattr(result, name)
+
+    return result
+
+
+def get_class(module: str, qualname: str):
+    key = module, qualname
+    try:
+        return TOKENIZABLE_CLASS_REGISTRY[key]
+    except KeyError:
+        cls = import_qualname(module, qualname, REMAPPED_CLASSES)
+        TOKENIZABLE_CLASS_REGISTRY[key] = cls
+        return cls
+
+
+def modify_dependencies(obj: Union[Dict, List], modifier, is_mine, top=True):
+    """
+    Parameters
+    ----------
+    obj : Dict or List
+        Dictionary to traverse. Assumes that only mappings are dict and only
+        iterables are list, and that no gufe objects are in the keys of
+        dicts
+    modifier : Callable[[GufeTokenizable], Any]
+        function that modifies any GufeTokenizable found
+    is_mine : Callable[Any, bool]
+        function that determines whether the given object should be
+        subjected to the modifier
+    top : bool
+        If `True`, skip modifying `obj` itself; needed for recursive use to
+        avoid early stopping on `obj`.
+    """
+    obj = copy.deepcopy(obj)
+
+    if is_mine(obj) and not top:
+        obj = modifier(obj)
+
+    if isinstance(obj, dict):
+        obj = {key: modify_dependencies(value, modifier, is_mine, top=False)
+               for key, value in obj.items()}
+
+    elif isinstance(obj, list):
+        obj = [modify_dependencies(item, modifier, is_mine, top=False)
+               for item in obj]
+
+    return obj
+
+
+# encode options
+def to_dict(obj: GufeTokenizable) -> Dict:
+    dct = obj._to_dict()
+    dct.update(module_qualname(obj))
+    return dct
+
+
+def dict_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        to_dict,
+        is_gufe_obj,
+        top=True
+    )
+
+
+def key_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        lambda obj: obj.key.to_dict(),
+        is_gufe_obj,
+        top=True
+    )
+
+
+# decode options
+def from_dict(dct) -> GufeTokenizable:
+    dct = copy.deepcopy(dct)
+
+    for key, val in dct.items():
+        if is_gufe_dict(val):
+            dct[key] = from_dict(val)
+
+    obj = _from_dict(dct)
+    try:
+        thing = TOKENIZABLE_REGISTRY[obj.key]()
+        # weakref will return None if the object was deleted
+        if thing is None:
+            return obj
+        else:
+            return thing
+    except KeyError:
+        return obj
```

Comment:
With the change I made in [ebb52e3](https://github.com/OpenFreeEnergy/gufe/pull/39/commits/ebb52e3bce1c7bdd4e753e794f6dda020bf6a33f), I think that this may now be impossible to reach: when your instantiate `obj`, it registers itself if something hasn't already taken the spot for its key, so either a previous `thing` has existed or `thing is obj`.

I think it is good defensive coding to leave it here, but I think we should mark this no-cover.

Reply:
I'm fine with removing it; agree with your assessment that it shouldn't be present. If we observe that there are edge cases in which we get a `KeyError`, then we'll need to understand and deal with them.